packages/react-components/react-utilities/src/compose/slot.ts

```diff
@@ -0,0 +1,117 @@
+import { isValidElement } from 'react';
+import type {
+  AsIntrinsicElement,
+  SlotComponent,
+  SlotComponentMetadata,
+  SlotRenderFunction,
+  SlotShorthandValue,
+  UnknownSlotProps,
+} from './types';
+import { SLOT_COMPONENT_METADATA_SYMBOL } from './constants';
+import * as React from 'react';
+import { SlotPropsRecord } from './types';
+import { ExtractSlotProps } from './types';
+import { PropsWithoutRef } from './types';
+import { getNativeElementProps } from '../utils/getNativeElementProps';
+
+export type SlotOptions<Props extends UnknownSlotProps> = {
+  elementType?:
+    | React.ComponentType<Props>
+    | (Props extends AsIntrinsicElement<infer As> ? As : keyof JSX.IntrinsicElements);
+  defaultProps?: Partial<Props>;
+};
+
+/**
+ * Creates a slot for a v9 component
+ * @param params - an object containing:
+ * 1. `value` provided through `props.SLOT`,
+ * 2. Base `elementType` that will be rendered
+ * 3. A boolean (`required`) indicating if the slot will be rendered even if no `value` is provided
+ * 4. `defaultProps`, similar to a component a slot might have default properties to be merged together with the value provided by the user
+ *
+ * @example
+ * ```ts
+ * // useDialogSurface
+ * const backdropSlot = slot({
+    value: props.backdrop,
+    elementType: 'div',
+    required: open && modalType !== 'non-modal',
+    defaultProps: { 'aria-hidden': 'true' },
+  });
+ * ```
+ */
+export function slotFromShorthand<Props extends UnknownSlotProps>(
+  value: Props | SlotShorthandValue | undefined,
+  options: { required: true } & SlotOptions<Props>,
+): Props;
+export function slotFromShorthand<Props extends UnknownSlotProps>(
+  value: Props | SlotShorthandValue | undefined | null,
+  options?: { required?: boolean } & SlotOptions<Props>,
+): Props | undefined;
+export function slotFromShorthand<Props extends UnknownSlotProps>(
+  value: Props | SlotShorthandValue | undefined | null,
+  options: { required?: boolean } & SlotOptions<Props> = {},
+): Props | undefined {
```

Comment:
Is the plan still to give these different names? `slot.required` or `requiredSlot`, etc.?